### DISCUSSION TITLE
How can I step over simplification?

### DISCUSSION THREAD

**Discussion by ghost** [josdejong/mathjs#2893] [Q&A]
Hi,

I now found this lib and I think it really cool, so thanks for the project.
My doubt is just: how can I step on the simplification?

For exemple:
take a algebraic expression like: (a + 2a) /(a + 2a)*4 + 8 / 2
What I want to is: 
1-use the simplify but with receive some notification that the (a +2a) was divided, like a event or even something likes. Is possible? How?
2-or, instead of 1 above, step each call to simplify soomething like this:
input = (a + 2a) /(a + 2a)*4 + 8 / 2
call 1 = 1 / 4 + 8 / 2
call 2 = 1 / 4 + 4
call 3 = 17 / 4
Each call happens only when I run the function, and so, I can register the steps, something like put some "explain" about it.

Some ideia on that?

**Comment by josdejong** [comment]
Thanks!

The `simplify` function comes with a list of built-in rules (`math.simplify.rules`). You can pass your own list of rules as second argument, like `simplify(expr, rules)`. If you want control over every step, you can write a loop yourself, pass only a single rule every time you execute `simplify`, and take the result as input for the second rule, etc.

**Reply by ghost** [reply]
> Thanks!
> 
> The `simplify` function comes with a list of built-in rules (`math.simplify.rules`). You can pass your own list of rules as second argument, like `simplify(expr, rules)`. If you want control over every step, you can write a loop yourself, pass only a single rule every time you execute `simplify`, and take the result as input for the second rule, etc.

I see. Thank you, I will search on the docs. Yesterday, I didn't tested the lib by myself, but just look over the examples. I will take a slow look at the docs and try by myself.
